### PR TITLE
Add C#-support via Roslyn.

### PR DESCRIPTION
**Pull request by josteink**
I've tested it and it seems to work. OOB it makes a few assumptions:

* Unix-platform where OmniSharp is executed via mono
* OmniSharp-Roslyn unpacked to ~/.omnisharp

This can be customized by the user if they prefer something else.

This fixes https://github.com/emacs-lsp/lsp-mode/issues/815.

### CODE REVIEW

**Comment by josteink** [comment]
Updated to automatically detect need for mono, and only customisation needed is path to OmniSharp-Roslyn.exe-file.

File: lsp-clients.el [hunk 1]
```diff
@@ -511,6 +511,30 @@ responsiveness at the cost of possibile stability issues."
...
+(defcustom lsp-clients-csharp-language-server-path
+  (expand-file-name "~/.omnisharp/omnisharp/omnisharp/OmniSharp.exe")
+  "The path to the OmnisSharp Roslyn language-server."
+  :group 'lsp-csharp
+  :type '(string :tag "Single string value"))
+
+(defun lsp-clients-csharp-language-server-command ()
+  (if (eq system-type 'windows-nt)
+      (list lsp-clients-csharp-language-server-path "-lsp")
+    (list "mono" lsp-clients-csharp-langauge-server-path "-lsp")))
```
Comment (yyoncho) [inline lsp-clients.el]: you have a typo

Comment (josteink) [reply lsp-clients.el]: Oh indeed. Will fix!

File: lsp-clients.el [hunk 2]
```diff
@@ -511,6 +511,30 @@ responsiveness at the cost of possibile stability issues."
...
                   :server-id 'angular-ls))
 
+;;; C-sharp
+(defgroup lsp-csharp nil
+  "LSP support for C#, using the Omnisharp Language Server. Version 1.34.3 minimum required."
+  :group 'lsp-mode
+  :link '(url-link "https://github.com/OmniSharp/omnisharp-roslyn"))
+
+(defcustom lsp-clients-csharp-language-server-path
+  (expand-file-name "~/.omnisharp/omnisharp/omnisharp/OmniSharp.exe")
```
Comment (yyoncho) [inline lsp-clients.el]: is 3 times omnisharp intentional?

Comment (josteink) [reply lsp-clients.el]: If you unpack the OmniSharp Roslyn tarball in ‘~/.omnisharp/‘, this is what you get, yes.

**Comment by yyoncho** [comment]
Can you help with with picking up the correct version? I downloaded `1.34.4-beta.3` using https://roslynomnisharp.blob.core.windows.net/releases?restype=container&comp=list and then I created a new project using `dotnet new console` and when I opened the project it connected correctly but neither the completion or errors were working properly.

**Comment by josteink** [comment]
I downloaded this release straight off GitHub: https://github.com/OmniSharp/omnisharp-roslyn/releases/tag/v1.34.3

It worked on a few “traditional” .NET framework projects I have lying around. Didn’t test with .NET Core to be honest.

**Comment by josteink** [comment]
That said, if getting it to work isn’t easily reproducible, it might need some more work.

**Comment by yyoncho** [comment]
> That said, if getting it to work isn’t easily reproducible, it might need some more work.

It might me my fault, not having what is required for normal C# development under linux. 

> It worked on a few “traditional” .NET framework projects I have lying around. Didn’t test with .NET Core to be honest.

Can you link me such a project? Sorry for asking but I am not very familiar with dotnet trends. Also, on linux with mono, which archieve should I use omnisharp-mono.zip or omnisharp-linux-x86.tar.gz ?

**Comment by josteink** [comment]
> Also, on linux with mono, which archieve should I use omnisharp-mono.zip or omnisharp-linux-x86.tar.gz ?

The latter should be fine. It’s what I use. 

I also have mono-complete installed on my Ubuntu 19.04 installation.

**Comment by josteink** [comment]
> Can you link me such a project? Sorry for asking but I am not very familiar with dotnet trends.

I've done some more testing, and I must admit I'm having around 50/50 success rate with various projects. Some projects (.NET Framework & .NET Core) report `System.NullReferenceException` in *lsp-log*, and others don't.

I really think that is an issue with OmniSharp, more so than patch per. That said: if OmniSharp cannot yet reliably work with all C#, solutions, I guess you need to decide if you want to merge this or not.

**Comment by yyoncho** [comment]
> I really think that is an issue with OmniSharp, more so than patch per. That said: if OmniSharp cannot yet reliably work with all C#, solutions, I guess you need to decide if you want to merge this or not.

Can you send me a project that work or give me instructions how to make such project? IMO we should merge it and add comment in the readme about the state of the language server and then start reporting upstream issues.

**Comment by yyoncho** [comment]
As a side note, I am adding the omnisharp server to the docker image from defined here https://github.com/emacs-lsp/lsp-docker .

**Comment by yyoncho** [comment]
I pushed https://github.com/emacs-lsp/lsp-docker/commit/2143b7b2bf06b1c6088e5db7e5a2a02382d098ec and I am having the same result for the demo project in the PR.

**Comment by josteink** [comment]
Thanks for taking your time to properly test and review this. I'll try to get back to your tomorrow with some demo-projects where we can hopefully both reproduce the same (expected) behaviour.

That said, I see that the project you added to ̀lsp-docker` is the "new" kind of project associated with .NET Core, not traditional .NET framework.

I'll let you know if this affects things one way or the other when I've done some more cross-framework testing.

**Comment by josteink** [comment]
An example project where LSP-mode seems to work reasonably fine is "workspacer", which is written in "old style" .NET project files: https://github.com/rickbutton/workspacer/

Not sure why it seems to work better with some projects and not with others yet. Haven't had any time to test and look into it deeper.

**Comment by yyoncho** [comment]
Thank you and congrats for your first `lsp-mode` PR! :tada:

**Pull request merged**